lucene/core/src/java/org/apache/lucene/search/SortField.java

```diff
@@ -120,6 +126,104 @@ public SortField(String field, Type type, boolean reverse) {
     this.reverse = reverse;
   }
 
+  /** A SortFieldProvider for field sorts */
+  public static final class Provider extends SortFieldProvider {
+
+    /** The name this Provider is registered under */
+    public static final String NAME = "SortField";
+
+    /** Creates a new Provider */
+    public Provider() {
+      super(NAME);
+    }
+
+    @Override
+    public SortField readSortField(DataInput in) throws IOException {
+      SortField sf = new SortField(in.readString(), readType(in), in.readInt() == 1);
+      if (in.readInt() == 1) {
+        // missing object
+        switch (sf.type) {
+          case STRING:
+            int missingString = in.readInt();
+            if (missingString == 1) {
+              sf.setMissingValue(STRING_FIRST);
+            }
+            else {
+              sf.setMissingValue(STRING_LAST);
+            }
+            break;
+          case INT:
+            sf.setMissingValue(in.readInt());
+            break;
+          case LONG:
+            sf.setMissingValue(in.readLong());
+            break;
+          case FLOAT:
+            sf.setMissingValue(NumericUtils.sortableIntToFloat(in.readInt()));
+            break;
+          case DOUBLE:
+            sf.setMissingValue(NumericUtils.sortableLongToDouble(in.readLong()));
+            break;
+          default:
+            throw new IllegalArgumentException("Cannot deserialize sort of type " + sf.type);
+        }
+      }
+      return sf;
+    }
+
+    @Override
+    public void writeSortField(SortField sf, DataOutput out) throws IOException {
+      sf.serialize(out);
+    }
+  }
+
+  protected static Type readType(DataInput in) throws IOException {
+    int type = in.readInt();
+    if (type >= Type.values().length) {
+      throw new IllegalArgumentException("Can't deserialize SortField - unknown type " + type);
+    }
+    return Type.values()[type];
+  }
+
+  private void serialize(DataOutput out) throws IOException {
+    out.writeString(field);
+    out.writeInt(type.ordinal());
```

Comment:
Can we look into not using enum ordinals for serialization? It makes something that looks as harmless as reordering an enum cause backward-compatibility breaks. E.g. maybe we could have a mapping between Type constants and int IDs as a constant on this class?